L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,462 @@
+
+One common requirement is to skip the hostname verification check if a client has a different way of validating the server's identity in different platforms. 
+For example, it's becoming more and more popular to use SPIFFE ID as the identity format, but a valid certificate using SPIFFE ID doesn't necessarily contain the dnsName in the SAN field, which might cause hostname verification to fail.
+
+Another requirement is to reload the root and identity credentials without shutting down the server, because certificates will eventually expire and the ability to rotate the certificates is essential to a cloud application.
+
+Beisdes requests from Open Source Community, with the growth of GCP, we've seen more and more internal use cases of these features.   
+
+Hence we propose an API that will meet the following requirements:  
+
+1. Flexible enough to support multiple use cases. Users should be able to write their own reloading or authorization logic if the provided ones doesn't fit their needs
+2. The credentials reloading should be efficient. We shall reload only when needed rather than reloading per TLS connection
+3. C core API needs to be simple enough, so that most of existing SSL credentials in wrapped languages can migrate to call this new API
+
+### Related Proposals:
+
+[L46: Create a new TLS credential API that supports SPIFFE mutual TLS](https://github.com/grpc/proposal/pull/98)
+
+## Proposal
+
+### TLS credentials and TLS Options
+This part of the proposal introduces `grpc_tls_credentials_options` and TLS credentials. Users use `grpc_tls_credentials_options` to configure the specific security properties they want, and build the client/server credential by passing `grpc_tls_credentials_options` in.
+This section only covers the configuration not related to credential reloading and custom verification. Configurations related to those two topics are explained in the latter sections.
+
+```c
+/** --- TLS channel/server credentials ---
+ * It is used for experimental purpose for now and subject to change. */
+
+typedef struct grpc_tls_credentials_options grpc_tls_credentials_options;
+
+GRPCAPI grpc_tls_credentials_options* grpc_tls_credentials_options_create(void);
+
+/**
+ * Set grpc_ssl_client_certificate_request_type in credentials options.
+ * If not set, we will use the default value GRPC_SSL_DONT_REQUEST_CLIENT_CERTIFICATE.
+ */
+GRPCAPI int grpc_tls_credentials_options_set_cert_request_type(
+    grpc_tls_credentials_options* options,
+    grpc_ssl_client_certificate_request_type type);
+
+/**
+ * Set grpc_tls_server_verification_option. 
+ * If not set, we will use the default value GRPC_TLS_SERVER_VERIFICATION.
+ * This should be called only on the client side.
+ */
+GRPCAPI int grpc_tls_credentials_options_set_server_verification_option(
+    grpc_tls_credentials_options* options,
+    grpc_tls_server_verification_option server_verification_option);
```

Comment:
Do you plan to introduce `grpc_tls_client_verification_option` enum for the client-side custom authorization check?

Reply:
I can do either way. Having a `grpc_tls_client_verification_option` is what we are doing in Go.

If server side doesn't want to verify the certificate, it can choose the corresponding `GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_BUT_DONT_VERIFY` value in `grpc_ssl_client_certificate_request_type`, which I think is equivalent here. But I am also fine adding a `grpc_tls_client_verification_option` to make it (probably) easier for users to understand, etc.

Reply:
+1 on having `grpc_tls_client_verification_option` for extendability.